aten/src/ATen/core/ivalue_inl.h

```diff
@@ -412,6 +412,13 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   c10::optional<FutureError> error_;
 };
 
+// Input is a list of Futures with the same target type.
+// Output is a Future to the List of completed Futures.
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
```

Comment:
Ok it seems like this is where we want the c10::List to have ivalue::Future, but why we have two APIs here? shouldn't the second one be enough?

Reply:
My intent was to just create a convenience api (just like some of the alternate constructor apis in ivalue.h), since List<> can entail some effort to use (though maybe just me? :) ). I can remove if we'd really like.

Reply:
Actually, looking at this a bit more, I'll remove this secondary api.

Reply:
Actually, un-removing after hacking on the unittest a bit, unless there's an objection.

Reply:
How would this c++ API getting used in the rpc? I don't have strong opinion on if we should use c10::List as a C++ API for this, but one drawback I am seeing here is, c10::List is suppose to holding a GenericList with the same type (acting like a real list that contains a single type), but if we expose this c10::List with Future as a exception on the `static_assert` we discussed above, we might ended up having something like `List[Future[int], Future[double]]`, which is not desirable as it break the list contraints in the type system, which might potentially cause some issue.

Reply:
Re: rpc - since rpc exported api is based on `torch::util::Future<Message>` (since `Message` isn't an `IValue`), it should be straightforward to create a parallel version for the `torch::util::Future` type.

Re: `c10::List<>` - regardless of the input (`vector` vs `List`), we need to output a `c10::List`, because it needs to be the object of a `Future<>` (i.e. `Future<List<Future<Any>>>`. If we use `type FutureType(AnyType)` in the `c10::List`, things do seem to work (and not like `c10::List` enforces this for `Future` - remember the lack of the `call` operator, or type firmly embedded in the template)

Reply:
What I am trying to see for this API is that, are we having the case where we want `collectAll` takes a list like `List[Future[int], Future[double], Future[Tensor]`? Or should `collectAll` be only applies to a list that have the same type? 

If the use case is the later one, we should just use the `std::vector` for all cases (c10::List version can be used internally as implementation to generate the new Future)? at least this seems to be the one used by C++ tests natively. We should also create the `Future<List<Future<TypeFromInput>>>` instead of AnyType to make the type consistent before and after.